// react-native-css-stylesheet/lib/commonjs/calculations.d.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.viewportWidth = exports.viewportMin = exports.viewportMax = exports.viewportHeight = exports.verticalScale = exports.sizeRegex = exports.setGuidelineBaseDimensions = exports.setDefaultFontSize = exports.scale = exports.rem = exports.moderateVerticalScale = exports.moderateScale = exports.isValidSize = exports.getCurrentOrientation = exports.calculateSize = void 0;
const sizeRegex;
exports.sizeRegex = sizeRegex;
const scale;
exports.scale = scale;
const verticalScale;
exports.verticalScale = verticalScale;
const moderateScale;
exports.moderateScale = moderateScale;
const moderateVerticalScale;
exports.moderateVerticalScale = moderateVerticalScale;
const rem;
exports.rem = rem;
const viewportMin;
exports.viewportMin = viewportMin;
const viewportMax;
exports.viewportMax = viewportMax;
const viewportHeight;
exports.viewportHeight = viewportHeight;
const viewportWidth;
exports.viewportWidth = viewportWidth;
const isValidSize;
exports.isValidSize = isValidSize;
const calculateSize;
exports.calculateSize = calculateSize;
const getCurrentOrientation;
exports.getCurrentOrientation = getCurrentOrientation;
const setGuidelineBaseDimensions;
exports.setGuidelineBaseDimensions = setGuidelineBaseDimensions;
const setDefaultFontSize;
exports.setDefaultFontSize = setDefaultFontSize;
